cron/worker/main.go

```diff
@@ -0,0 +1,136 @@
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/google/go-github/v32/github"
+	"github.com/ossf/scorecard/checks"
+	"github.com/ossf/scorecard/cron/config"
+	"github.com/ossf/scorecard/cron/data"
+	"github.com/ossf/scorecard/cron/pubsub"
+	"github.com/ossf/scorecard/pkg"
+	"github.com/ossf/scorecard/repos"
+	"github.com/ossf/scorecard/roundtripper"
+	"github.com/shurcooL/githubv4"
+	"go.uber.org/zap"
+)
+
+func processRequest(ctx context.Context,
+	batchRequest *data.ScorecardBatchRequest, bucketURL string,
+	httpClient *http.Client, githubClient *github.Client, graphClient *githubv4.Client) error {
+	repoURLs := make([]repos.RepoURL, len(batchRequest.GetRepos()))
+	var resultCh chan repos.RepoResult
+	var buffer bytes.Buffer
+	checksToRun := checks.AllChecks
+
+	for _, repo := range batchRequest.GetRepos() {
+		repoURL := repos.RepoURL{}
+		if err := repoURL.Set(repo); err != nil {
+			return fmt.Errorf("error setting RepoURL: %w", err)
+		}
+		if err := repoURL.ValidGitHubURL(); err != nil {
+			return fmt.Errorf("url is not a valid GitHub URL: %w", err)
+		}
+		repoURLs = append(repoURLs, repoURL)
+	}
+	go func() {
+		var wg sync.WaitGroup
+		for _, repoURL := range repoURLs {
+			repoURL := repoURL
+			wg.Add(1)
+			go func() {
+				resultCh <- pkg.RunScorecards(ctx, repoURL, checksToRun, httpClient, githubClient, graphClient)
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+	}()
+
+	for result := range resultCh {
+		err := result.AsJSON(true /*showDetails*/, &buffer)
+		if err != nil {
+			return fmt.Errorf("error during result.AsJSON: %w", err)
+		}
+	}
+
+	filename := data.GetBlobFilename(
+		fmt.Sprintf("shard-%03d", batchRequest.GetShardNum()),
+		batchRequest.GetJobTime().AsTime())
+	return errors.Unwrap(data.WriteToBlobStore(ctx, bucketURL, filename, buffer.Bytes()))
+}
+
+func createNetClients(ctx context.Context) (httpClient *http.Client,
+	githubClient *github.Client, graphClient *githubv4.Client) {
+	cfg := zap.NewProductionConfig()
+	cfg.Level.SetLevel(zap.InfoLevel)
+	logger, err := cfg.Build()
+	if err != nil {
+		panic(err)
+	}
+	sugar := logger.Sugar()
+	// Use our custom roundtripper
+	rt := roundtripper.NewTransport(ctx, sugar)
+	httpClient = &http.Client{
+		Transport: rt,
+	}
+	githubClient = github.NewClient(httpClient)
+	graphClient = githubv4.NewClient(httpClient)
+	return
+}
+
+func main() {
+	ctx := context.Background()
+
+	subscriptionURL, err := config.GetRequestSubscriptionURL()
+	if err != nil {
+		panic(err)
+	}
+	subscriber, err := pubsub.CreateSubscriber(ctx, subscriptionURL)
+	if err != nil {
+		panic(err)
+	}
+
+	bucketURL, err := config.GetResultDataBucketURL()
+	if err != nil {
+		panic(err)
+	}
+
+	httpClient, githubClient, graphClient := createNetClients(ctx)
+
+	for {
+		req, err := subscriber.SynchronousPull()
+		if errors.Is(err, pubsub.ErrorInReceive) {
+			fmt.Printf("a non-retryable error has occurred during Receive, exiting: %v", err)
+			break
+		} else if err != nil {
+			panic(err)
```

Comment:
Why do we need to distinguish between this case and line 122? They ultimately both cause us to exit right?

Reply:
This line is a error in our message handling (e.g: JSON parsing error). It means there is something wrong in our code, so we explicitly `panic`.

An error during Receive (line 122) is from the PubSub driver itself. It could be a harmless error (e.g: no messages to poll) and will exit without `panic`.

Reply:
> I see, thanks. I think this does make things a bit complicated though, because it puts the burden on the caller to know about this detail (and it's not really an error).
> 
> Can we just return `req == nil` and `err == nil` in this case and handle it here accordingly? 

Done.

>Also should we exit here or just sleep for a bit and retry later?

Quitting the program here is what gocloud recommends. But considering we'll be updating from go-cloud to GCS, this may not hold true anymore. 
 
![image](https://user-images.githubusercontent.com/4825078/118712837-19488700-b7d6-11eb-8270-7f3370ca7c50.png)

Reply:
Right, we could have also recreated the subscription instead of exiting, but yes no longer necessary for the google cloud libraries.